internal/pki/pki.go

```diff
@@ -194,9 +197,24 @@ func (root *RootCertificateAuthority) leafIsValid(leaf *LeafCertificate) bool {
 		leaf.PrivateKey.ecdsa != nil &&
 		leaf.PrivateKey.ecdsa.PublicKey.Equal(leaf.Certificate.x509.PublicKey)
 
+	// It is not yet past the "renewal by" time,
+	// as defined by the before and after times of the certificate's expiration
+	// and the default ratio
+	ok = ok && isBeforeRenewalTime(leaf.Certificate.x509.NotBefore,
+		leaf.Certificate.x509.NotAfter)
+
 	return ok
 }
 
+// isBeforeRenewalTime checks if the result of `currentTime`
+// is after the default renewal time of
+// 1/3rds before the certificate's expiry
+func isBeforeRenewalTime(before, after time.Time) bool {
+	renewalDuration := after.Sub(before) / renewalRatio
+	renewalTime := after.Add(-1 * renewalDuration)
+	return currentTime().Before(renewalTime)
```

Comment:
🌱 Knowing that we read from a cache, I wonder if we should do something somewhere to avoid generating certificates when the cache is stale.

Have you ever seen a leaf get generated "extra" times while testing?

Reply:
I have not seen extra generating of the leaf cert, but here's the situation where I didn't see one: 

I had a print line in here at one time, and a setup where most of my certs were year-long, and one cert was valid for 3 hours (starting an hour ago, so I expected a regen in one hour); when that hour was up and I had triggered a reconcile, I only saw this give a `false` value once for that one cert.

This does make me want to put in an event when regenerating certs though, so that we could see if/when certs get regenerated multiple times before the cache is updated.